crates/shared/src/price_estimation/native/coingecko.rs

```diff
@@ -111,46 +128,93 @@ impl CoinGecko {
+        prices
             .get(&token)
-            .ok_or(PriceEstimationError::NoLiquidity)?;
-        Ok(price.eth)
+            .ok_or(PriceEstimationError::NoLiquidity)?
+            .clone()
+    }
+}
+
+#[async_trait]
+impl NativePriceBatchFetcher for CoinGecko {
+    async fn fetch_native_prices(
+        &self,
+        tokens: &HashSet<H160>,
+    ) -> std::result::Result<HashMap<H160, NativePriceEstimateResult>, PriceEstimationError> {
+        let mut requested_tokens = tokens.iter().collect::<Vec<_>>();
+        match self.quote_token {
+            QuoteToken::Eth => self.send_bulk_request_price_in_eth(&requested_tokens).await,
+            QuoteToken::Other(native_price_token) => {
+                requested_tokens.push(&native_price_token);
+                let prices = self
+                    .send_bulk_request_price_in_eth(&requested_tokens)
+                    .await?;
+                let native_price_token = prices
+                    .get(&native_price_token)
+                    .ok_or(PriceEstimationError::NoLiquidity)?
+                    .clone();
+                let native_price_token_eth = rust_decimal::Decimal::try_from(native_price_token?)
+                    .map_err(|e| {
+                    PriceEstimationError::EstimatorInternal(anyhow!(
+                        "failed to parse native price token in ETH to rust decimal: {e:?}"
+                    ))
+                })?;
+                prices
+                    .into_iter()
+                    .filter(|(token, _)| tokens.contains(token))
+                    .map(|(token, price)| match price.as_ref() {
+                        Ok(price) => {
+                            let token_eth =
+                                rust_decimal::Decimal::try_from(*price).map_err(|e| {
+                                    PriceEstimationError::EstimatorInternal(anyhow!(
+                                        "failed to parse requested token in ETH to rust decimal: \
+                                         {e:?}"
+                                    ))
+                                })?;
+                            let token_in_native_price = token_eth
+                                .checked_div(native_price_token_eth)
+                                .ok_or(PriceEstimationError::EstimatorInternal(anyhow!(
+                                    "division by zero"
+                                )))?;
+                            let token_in_native_price_f64 = token_in_native_price.to_f64().ok_or(
+                                PriceEstimationError::EstimatorInternal(anyhow!(
+                                    "failed to parse result to f64"
+                                )),
+                            )?;
+                            Ok((token, Ok(token_in_native_price_f64)))
+                        }
+                        Err(_) => Ok((token, price)),
+                    })
+                    .collect::<Result<_, _>>()
```

Comment:
Collecting into a single `Result` means a single result will invalidate the whole batch. Only omitting the bad prices and returning all the good ones seems like the better choice.